File: helpers/templates/python/benchmark.py
"""Timing benchmark for {{TITLE}}.

Run with: make bench DIR=problems/<platform>/<slug>   (or: python benchmark.py)
"""

import statistics
import time

from solution import solve

SIZES = [10, 1_000, 100_000]  # input sizes to measure (lower them for slow solutions)
REPEATS = 7  # timed runs per size; the minimum and the median are reported


def make_input(n):
    """Return the arguments for solve() as a tuple, for an input of size n.

    TODO: build a realistic (ideally worst-case) input, e.g. `return (list(range(n)), n)`.
    If solve() changes its input, build a fresh copy inside measure() for every run.
    """
    return ()


def measure(n):
    args = make_input(n)
    solve(*args)  # warmup
    times = []
    for _ in range(REPEATS):
        start = time.perf_counter()
        solve(*args)
        times.append((time.perf_counter() - start) * 1000)
    return min(times), statistics.median(times)


if __name__ == "__main__":
    print(f"{'size':<12}{'min ms':>12}{'median ms':>14}")
    for n in SIZES:
        fastest, median = measure(n)
        print(f"{n:<12}{fastest:>12.6f}{median:>14.6f}")

File: helpers/templates/rust/benchmark.rs
//! Timing benchmark for {{TITLE}}.
//! Run with: make bench DIR=problems/<platform>/<slug>

#[allow(dead_code)]
#[path = "solution.rs"]
mod solution;

use std::hint::black_box;
use std::time::Instant;

/// Input sizes to measure (lower them for slow solutions).
const SIZES: [usize; 3] = [10, 1_000, 100_000];
/// Timed runs per size; the minimum and the median are reported.
const REPEATS: usize = 7;

fn main() {
    println!("{:<12}{:>12}{:>14}", "size", "min ms", "median ms");
    for &n in &SIZES {
        // TODO: build the input for size n (ideally worst case) once, here, and pass it to
        // solve(...) below. If solve() changes its input, copy it inside the timed loop.
        let _ = n;
        black_box(solution::solve()); // warmup
        let mut times = Vec::with_capacity(REPEATS);
        for _ in 0..REPEATS {
            let start = Instant::now();
            black_box(solution::solve());
            times.push(start.elapsed().as_secs_f64() * 1000.0);
        }
        times.sort_by(|a, b| a.partial_cmp(b).unwrap());
        println!("{:<12}{:>12.6}{:>14.6}", n, times[0], times[REPEATS / 2]);
    }
}
